### PR TITLE
Make SizeClasses as composition with PoolArena instead of inheritance to reduce memory usage

### PR DESCRIPTION
**Pull request by laosijikaichele**
Motivation:

For one `PooledByteBufAllocator` instance, the array `PooledByteBufAllocator.heapArenas[]` only needs one `SizeClasses` instance, so does the `PooledByteBufAllocator.directArenas[]`, this will reduce memory usage.

Take the `PooledByteBufAllocator.DEFAULT` as an example, on my laptop, after `PooledByteBufAllocator.DEFAULT` initialized,  there are `heapArenas[16]` and `directArenas[16]` arrays created, every element of the arrays created following `int[]` array by `SizeClasses`'s constructor:
```
int[] pageIdx2sizeTab; // Length is 32 on my laptop
int[] sizeIdx2sizeTab; // Length is 68 on my laptop
int[] size2idxTab; // Length is 256 on my laptop
```
After `PooledByteBufAllocator.DEFAULT` initialized, there are 32 `int[] pageIdx2sizeTab`, 32 `int[] sizeIdx2sizeTab`, and 32 `int[] size2idxTab` arrays created. 

This is not necessary, because the `PooledByteBufAllocator.heapArenas[]` can share one `SizeClasses` instance, so does the `PooledByteBufAllocator.directArenas[]`. 

After the PR change, the `PooledByteBufAllocator.DEFAULT` only needs 2 `int[] pageIdx2sizeTab`, 2 `int[] sizeIdx2sizeTab` and 2 `int[] size2idxTab` arrays. 

It reduced at least 42720 bytes memory usage when use `PooledByteBufAllocator.DEFAULT`:
4 * (30 * 32 + 30 * 68 + 30 * 256) = 42720 bytes. 

So we make `SizeClasses` as composition with `PoolArena` instead of inheritance.
And it still reserves the ability to create `SizeClasses` per `PoolArena` for possible different implementation in the future.

Modification:

Make `SizeClasses` as composition with `PoolArena` instead of inheritance.

Result:

Reduced memory footprint.

### CODE REVIEW

**Comment by laosijikaichele** [comment]
@normanmaurer done.

File: buffer/src/main/java/io/netty/buffer/SizeClasses.java [hunk 1]
```diff
@@ -78,14 +78,13 @@
  * <p>
  *   ( 76,    24,       22,        1,       yes,            no,        no)
  */
-abstract class SizeClasses implements SizeClassesMetric {
+class SizeClasses implements SizeClassesMetric {
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/SizeClasses.java:81]: Consider making it `final` to ensure we lock out inheritance.

**Review by chrisvest — APPROVED** [review]
No need to forward port this because `main` does a similar thing but differently.

File: buffer/src/main/java/io/netty/buffer/SizeClasses.java (hunk 1 above)
Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/SizeClasses.java]: Done

**Review by chrisvest — APPROVED** [review]

**Pull request merged**